### PR TITLE
Hsm 302

### PR DESCRIPTION
**Pull request by aveenismail**
Add check of return codes of OpenSSL function

### CODE REVIEW

File: common/util.c [hunk 1]
```diff
@@ -36,9 +36,9 @@ bool set_component(unsigned char *in_ptr, const BIGNUM *bn, int element_len) {
 
   memset(in_ptr, 0, (size_t)(element_len - real_len));
   in_ptr += element_len - real_len;
-  BN_bn2bin(bn, in_ptr);
-
-  return true;
+  // BN_bn2bin(bn, in_ptr);
+  // return true;
```
Comment (klali) [inline common/util.c:40]: I think we should get rid of the commented out code here

File: common/util.c [hunk 2]
```diff
@@ -88,6 +88,9 @@ bool read_ed25519_key(uint8_t *in, size_t in_len, uint8_t *out,
   int ret;
   BIO *b64 = BIO_new(BIO_f_base64());
   BIO *bio = BIO_new(BIO_s_mem());
+  if (b64 == NULL || bio == NULL) {
```
Comment (klali) [inline common/util.c:91]: this may leak memory if allocation of b64 succeeds but bio fails (unlikely, yes.)

File: common/util.c [hunk 3]
```diff
@@ -538,6 +544,9 @@ bool base64_decode(const char *in, uint8_t *out, size_t *len) {
...
```
Comment (klali) [inline common/util.c:547]: this can leak memory.

File: common/util.c [hunk 4]
```diff
@@ -569,6 +578,9 @@ bool write_file(const uint8_t *buf, size_t buf_len, FILE *fp, format_t format) {
 
     b64 = BIO_new(BIO_f_base64());
     bio = BIO_new(BIO_s_mem());
+    if (b64 == NULL || bio == NULL) {
```
Comment (klali) [inline common/util.c:581]: this can leak memory.

File: src/commands.c [hunk 5]
```diff
@@ -2445,6 +2498,10 @@ int yh_com_sign_ssh_certificate(yubihsm_context *ctx, Argument *argv,
 
   b64 = BIO_new(BIO_f_base64());
   bio = BIO_new(BIO_s_mem());
+  if (b64 == NULL || bio == NULL) {
```
Comment (klali) [inline src/commands.c:2501]: this can leak memory.

File: common/util.c [hunk 6]
```diff
@@ -86,8 +84,17 @@ bool read_ed25519_key(uint8_t *in, size_t in_len, uint8_t *out,
...
-  BIO *bio = BIO_new(BIO_s_mem());
+  BIO *b64 = NULL;
+  BIO *bio = NULL;
+
+  b64 = BIO_new(BIO_f_base64());
+  if (b64 == NULL) {
+    return false;
+  }
+  bio = BIO_new(BIO_s_mem());
+  if (bio == NULL) {
```
Comment (klali) [inline common/util.c:95]: this construct still leaks if the first operation succeeds and the second fails.

Comment (aveenismail) [reply common/util.c:95]: It should be sufficient to call Bio_free(b64) if the second operation fails, right?

Comment (klali) [reply common/util.c:95]: yes, I think so

Comment (aveenismail) [reply common/util.c:95]: Fixed

**Review by klali — APPROVED** [review]

**Comment by aveenismail** [comment]
Merged this with the command line

**Pull request closed without merging**